### PR TITLE
feat/remove tagged users

### PR DESCRIPTION
**Pull request by rishirishhh**
Date: 11 04 2024

Developer Name: Kavya Rishi

---

## Issue Ticket Number 
Removing users who satisfy certain conditions #229

## Description

There are certain users whose tasks go red as they have not completed the task in time or there are certain users who are tagged group-idle-7d+ . These users need to be removed from the discord server as they do not comply to the rules. Currently this task is done manually which needs to be done via a discord command
Users will be removed whenever they will match this condition

### Documentation Updated?

- [ ] Yes
- [X] No

### Under Feature Flag

- [ ] Yes
- [X] No

### Database Changes

- [ ] Yes
- [X] No

### Breaking Changes

- [ ] Yes
- [X] No

### Development Tested?

- [X] Yes
- [ ] No

## Screenshots
<details>

Screenshot from 2024-04-14 14-41-00

</details>

## Test Coverage
<details>

image

Screenshot from 2024-04-14 14-41-00

Screenshot from 2024-04-11 12-27-11

</details>

### CODE REVIEW

File: src/constants/commands.ts [hunk 1]
```diff
@@ -34,6 +34,19 @@ export const MENTION_EACH = {
   ],
 };
 
+export const KICK = {
+  name: "kick",
+  description: "Kick a user from the server",
```
Comment (iamitprakash) [inline src/constants/commands.ts:39]: use. Remove

File: src/constants/commands.ts [hunk 2]
```diff
@@ -34,6 +34,19 @@ export const MENTION_EACH = {
...
+  options: [
+    {
+      name: "role",
+      description: "The role to kick",
```
Comment (iamitprakash) [inline src/constants/commands.ts:43]: same here

File: src/controllers/kickEachUser.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  const getMembersInServerResponse = await getMembersInServer(env);
+  const roleId = transformedArgument.roleToBeRemovedObj.value;
+
+  const usersWithMatchingRole = filterUserByRoles(
+    getMembersInServerResponse as UserArray[],
+    roleId
+  );
+
+  if (usersWithMatchingRole.length === 0) {
+    return discordTextResponse(`Found no users with the matched role.`);
```
Comment (iamitprakash) [inline src/controllers/kickEachUser.ts:27]: need to re-phrase this message

File: src/controllers/kickEachUser.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  } else {
+    ctx.waitUntil(removeUsers(env, usersWithMatchingRole));
+    return discordTextResponse(
+      `Found ${usersWithMatchingRole.length} users with the matched role, removing them shortly...`
```
Comment (iamitprakash) [inline src/controllers/kickEachUser.ts:31]: don't we need name here

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: src/controllers/kickEachUser.ts (hunk 3 above)
Comment (iamitprakash) [reply src/controllers/kickEachUser.ts:27]: @rishirishhh  We couldn't find any users assigned to <role_name> role.

File: src/controllers/baseHandler.ts [hunk 5]
```diff
@@ -48,6 +50,8 @@ export async function baseHandler(
 ): Promise<JSONResponse> {
   const command = lowerCaseMessageCommands(message);
 
+  console.log("Message: ", JSON.stringify(message.data));
+  console.log("Envior:", env);
```
Comment (sahsisunny) [inline src/controllers/baseHandler.ts:54]: ?

Comment (rishirishhh) [reply src/controllers/baseHandler.ts]: This was added by mistake 
I have fixed it in my new commit

Comment (sahsisunny) [reply src/controllers/baseHandler.ts]: ok

File: src/utils/removeUsers.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import { env } from "../typeDefinitions/default.types";
+import { DISCORD_BASE_URL } from "../constants/urls";
+import {
+  parseRateLimitRemaining,
+  parseResetAfter,
+} from "./batchDiscordRequests";
+
+export async function removeUsers(env: env, usersWithMatchingRole: string[]) {
+  const batchSize = 4;
```
Comment (sahsisunny) [inline src/utils/removeUsers.ts:9]: I think is should be in constants.

Comment (rishirishhh) [reply src/utils/removeUsers.ts:9]: it is in constant

**Comment by sahsisunny** [comment]
> ## Description
> There are certain users whose tasks go red as they have not completed the task in time or there are certain users who are tagged group-idle-7d+ . These users need to be removed from the discord server as they do not comply to the rules. Currently this task is done manually which needs to be done via a discord command Users will be removed whenever they will match this condition
> 


This command removes the users who have specific roles right?

**Comment by rishirishhh** [comment]
> > ## Description
> > There are certain users whose tasks go red as they have not completed the task in time or there are certain users who are tagged group-idle-7d+ . These users need to be removed from the discord server as they do not comply to the rules. Currently this task is done manually which needs to be done via a discord command Users will be removed whenever they will match this condition
> 
> This command removes the users who have specific roles right?

Yes this command removes the users with specific roles

**Review by sahsisunny — DISMISSED** [review]
😢

**Comment by rishirishhh** [comment]
Screenshot from 2024-04-14 14-41-00
@ardourApeX 
this the updated result

**Review by ardourApeX — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**